### ISSUE TITLE
Entrypoint issue when volume mounting /usr/local/lib/node_modules

### ISSUE THREAD

**Issue by anjunatl** [jesseditson/absolutize-links#1]
This is almost definitely more of a docker-related question than anything else, but I figure since you've been through this you're probably the best person to ask -

When I have `- /usr/local/lib/node_modules:/usr/local/lib/node_modules` in the `volumes` section of my service in my `docker-compose.yml`, I get `stat /usr/local/bin/npm: no such file or directory` - presumably from the `entrypoint` command

```
services:
  npm: &npm
    build: dev-env
    volumes:
      - .:/opt/project
      - /usr/local/lib/node_modules:/usr/local/lib/node_modules
    working_dir: /opt/project/app
    entrypoint: "/usr/local/bin/npm"
```

There is already a `/usr/local/lib/node_modules` in the docker container, and `/usr/local/bin/npm` exists in the container, so I don't know why I'm getting that error.

```
root@c584d87f976a:/# cd /usr/local/lib/node_modules/
root@c584d87f976a:/usr/local/lib/node_modules# ls
gulp  npm
root@c584d87f976a:/usr/local/lib/node_modules# cd ../../bin/
root@c584d87f976a:/usr/local/bin# ls
gulp  node  nodejs  npm  yarn  yarnpkg
root@c584d87f976a:/usr/local/bin#
```

And on my host machine (Mac), I've got `/usr/local/lib/node_modules`

```
➜  node_modules pwd
/usr/local/lib/node_modules
```

I'm not sure why I'd be getting that, especially since the entrypoint is in `/usr/local/bin` and the volume mount is for `/usr/local/lib`.

Any help would be appreciated - thanks!

**Comment by jesseditson** [comment]
what does ls -al `/usr/local/bin` print? is `npm` symlinked to inside `/usr/local/lib/node_modules`?

One theory would be:
- `npm` is a symlink
- `/usr/local/lib/node_modules` is not available to docker, so it fails to link it to your container
- therefore, `/usr/local/bin/npm` is a symlink pointing to nothing

A fix for that scenario may be to make sure `/usr/local/lib/node_modules` is listed in the shared folders section of docker for mac preferences.

Another similar possibility, if `/usr/local/lib/node_modules` _is_ properly linked, but the symlink in `/usr/local/bin` is relative and points to nothing when in the context of your docker machine.

To get closer to the problem, try this:

`docker-compose run --rm npm bash`

That should get you an interactive shell to the container in the same state that is failing. Then, make sure `npm` is in fact in `/usr/local/bin`:

`ls -al /usr/local/bin`

If that prints a symlink, make sure the destination exists, or just run it: `/usr/local/bin/npm --version`.

Failing that, check if `npm` is installed & available: `which npm`. If not, maybe look at your $PATH: `echo $PATH`, and check that it includes `/usr/local/bin`. If that's set up, you should be able to change your entrypoint to just `npm`, although I'm not sure how much of a difference that would make.

Hopefully that should get you started with some ideas, LMK if that helps or if you still end up with issues.

**Comment by anjunatl** [comment]
```
> ls -al /usr/local/bin | grep npm
lrwxr-xr-x    1 anjunatl  admin        46 Jan 20  2016 npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
```

My Mac's npm is a symlink with an absolute path, from [node version manager](https://github.com/creationix/nvm). Due to the dev environment our large project is in, I'm not using Docker for Mac so I don't have the Docker for Mac screen you've mentioned.

```
root@c98ec7e556ba:/usr/local/bin# ls -la
total 29172
drwxrwxr-x  2  500   500     4096 Aug 10 00:39 .
drwxrwsr-x 21 root staff     4096 Aug 10 00:39 ..
lrwxrwxrwx  1 root root        36 Aug 10 00:39 gulp -> ../lib/node_modules/gulp/bin/gulp.js
-rwxrwxr-x  1  500   500 29862509 Aug  1 05:33 node
lrwxrwxrwx  1 root root        19 Aug  9 19:23 nodejs -> /usr/local/bin/node
lrwxrwxrwx  1  500   500       38 Aug  1 05:33 npm -> ../lib/node_modules/npm/bin/npm-cli.js
lrwxrwxrwx  1 root root        18 Aug  9 19:23 yarn -> /opt/yarn/bin/yarn
lrwxrwxrwx  1 root root        18 Aug  9 19:23 yarnpkg -> /opt/yarn/bin/yarn
root@c98ec7e556ba:/usr/local/bin# cd ../lib/node_modules/
root@c98ec7e556ba:/usr/local/lib/node_modules# ls -la
total 8
drwxr-sr-x 2 root staff 4096 Aug 14 14:41 .
drwxrwxr-x 5  500   500 4096 Aug 10 00:39 ..
root@c98ec7e556ba:/usr/local/lib/node_modules#
```

It appears the `/usr/local/lib/node_modules:/usr/local/lib/node_modules` volume mount inside the container from `docker-compose.yml` isn't working - but I've verified `/usr/local/lib/node_modules` exists on my Mac, is populated, and is not a symlink. I believe that's at least narrowed the problem down to the volume mount not working correctly - I'll play with it and see if I can come up with anything.

**Comment by anjunatl** [comment]
Figured it out. Our internal equivalent for the Docker for Mac (vm with a wrapper) only mounts our local `/Users/*` directory and has no access to local `/usr/` - Thanks for your time

**Comment by jesseditson** [comment]
Ah, thanks for following up! Yeah, the need to both mount and specify as a possible mount point to the VM is a confusing extra step. Before DFM I did this using a shared folder in virtualBox, if your VM software has a similar option, that should fix it for you.

**Comment by anjunatl** [comment]
I think that'd work for most cases, but our environment's complicated enough that it seems like there'd still be several steps involved for a dev to manually get it set up in our environment.

I tried mounting the folder of shareable components (the things we'd be `npm link`ing) into the container, since that lives in `/Users/blah/*` - then ran `npm link` from inside the container, and that appeared to work. I might even be able to set up some gulp watch tasks on the newly mounted folder to have it build when it notices changes in those files.

I'm still playing with it when I've got time though - I'm trying to find most efficient/least intrusive way of getting this working - the less steps/commands/edits my team will have to do to get this working, the better 😄
